File: web/tests/reference/intervals/employee-free-time.py
def employee_free_time(schedule):
    all_intervals = [iv for employee in schedule for iv in employee]
    all_intervals.sort(key=lambda iv: iv[0])

    merged = []
    for start, end in all_intervals:
        if merged and start <= merged[-1][1]:
            merged[-1][1] = max(merged[-1][1], end)
        else:
            merged.append([start, end])

    free_time = []
    for i in range(1, len(merged)):
        prev_end = merged[i - 1][1]
        curr_start = merged[i][0]
        if prev_end < curr_start:
            free_time.append([prev_end, curr_start])

    return free_time

File: web/tests/reference/intervals/employee-free-time.js
function employeeFreeTime(schedule) {
  const allIntervals = schedule.flat();
  allIntervals.sort((a, b) => a[0] - b[0]);

  const merged = [];
  for (const [start, end] of allIntervals) {
    if (merged.length && start <= merged[merged.length - 1][1]) {
      merged[merged.length - 1][1] = Math.max(merged[merged.length - 1][1], end);
    } else {
      merged.push([start, end]);
    }
  }

  const freeTime = [];
  for (let i = 1; i < merged.length; i++) {
    const prevEnd = merged[i - 1][1];
    const currStart = merged[i][0];
    if (prevEnd < currStart) freeTime.push([prevEnd, currStart]);
  }

  return freeTime;
}
